File: Two Pointers/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, next=head) # In case we need to remove the head node in the case of [1, 2], n = 2
        first, second = dummy, dummy
        
        for i in range(n):
            second = second.next
        
        while second.next:
            first, second = first.next, second.next
        
        first.next = first.next.next # First element is the element before the node that has to be removed
        
        return dummy.next

File: Two Pointers/19. Remove Nth Node From End of List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode *dummy = (struct ListNode*)malloc(sizeof(struct ListNode)*30);
    dummy->next = head;
    struct ListNode *ptr1 = dummy;
    struct ListNode *ptr2 = dummy;
    int i = 0;
    
    for (i=0; i<n; i++){
        ptr2 = ptr2->next;
    }
    
    while (ptr2->next != NULL){
        ptr1 = ptr1->next;
        ptr2 = ptr2->next;
    }
    
    ptr1->next = ptr1->next->next;
    
    return dummy->next;
}
